pyradiosky/skymodel.py

```diff
@@ -635,20 +899,59 @@ def skymodel_to_array(sky):
         recarray to turn into a SkyModel object.
 
     """
-    fieldtypes = ["U10", "f8", "f8", "f8", "f8"]
-    fieldnames = ["source_id", "ra_j2000", "dec_j2000", "flux_density_I", "frequency"]
-    fieldshapes = [()] * 3 + [(sky.Nfreqs,)] * 2
+    sky.check()
+    max_name_len = np.max([len(name) for name in sky.name])
+    fieldtypes = ["U" + str(max_name_len), "f8", "f8", "f8"]
+    fieldnames = ["source_id", "ra_j2000", "dec_j2000", "flux_density_I"]
+    fieldshapes = [()] * 3
+    if sky.freq_array is not None:
+        if sky.spectral_type == "subband":
+            fieldnames.append("subband_frequency")
+        else:
+            fieldnames.append("frequency")
+        fieldtypes.append("f8")
+        fieldshapes.extend([(sky.Nfreqs,)] * 2)
+    elif sky.reference_frequency is not None:
+        # add frequency field (a copy of reference_frequency) for backwards
+        # compatibility.
+        warnings.warn(
+            "The frequency field is included in the recarray as a copy of the "
+            "reference_frequency field for backwards compatibility. In future "
+            "only the reference_frequency will be included.",
+            category=DeprecationWarning,
+        )
```

Comment:
No, this is about using a field called `frequency` for both a `freq_array` and a `reference_frequency` -- a situation which is very confusing because they have different shapes and is hard to code properly for. I want to deprecate using the `frequency` field in favor of a `reference_frequency` column when we have a `reference_frequency` and not a `freq_array` on the object.

Reply:
So in the case of a `freq_array`, there simply won't be a frequency column at all, and the column headers for fluxes will indicate which frequency they correspond with?

Also, if we want to move from the column being called Frequency to being called reference_frequency, then the read_text_catalog method should also be updated to recognize either. The first generation reference simulations and a couple test files in `/data` label the column Frequency, but (conveniently) none of those Frequency values were ever used for anything. If the read method is made to ignore unrecognized columns, it should be fine going forward.